docs/zkapps/o1js/bitwise-operations.mdx

```diff
@@ -0,0 +1,345 @@
+hide_title: true
+description: Bitwise operations in o1js are implemented as gadgets.
+keywords:
+  - gadgets
+  - bitwise operations
+  - o1js
+---
+
+:::info
+
+Please note that zkApp programmability is not yet available on Mina Mainnet, but zkApps can now be deployed to Berkeley Testnet.
+
+:::
+
+# Bitwise Operations
+
+Bitwise operations manipulate individual bits within a binary represenbtation of a number. They can at times resemble boolean operations but will apply to a sequence of bits instead of booleans. Bitwise operations are generally available in most programming languages, including TypeScript. o1js provides versions of them that operate on `Field` elements and result in the necessary circuit constraints to generate a zero-knowledge proof of the computation. This is especially useful when implementing hashing algorithms such as SHA256.
+
+In o1js, bitwise operations and their attendant helper functions are implemented as [gadgets](/zkapps/o1js/gadgets).
+
+Bitwise operations:
+- [and()](#and)
+- [not()](#not)
+- [xor()](#xor)
+- [leftShift32()](#leftshift32)
+- [leftShift64()](#leftshift64)
+- [rightShift64()](#rightshift64)
+- [rotate32()](#rotate32)
+- [rotate64()](#rotate64)
+
+Helper functions:
+- [addMod32()](#addmod32)
+- [divMod32()](#divmod32)
+- [rangeCheck32()](#rangecheck32)
+- [rangeCheck64()](#rangecheck64)
+- [multiRangeCheck()](#multirangecheck)
+- [compactMultiRangeCheck()](#compactmultirangecheck)
+
+## and()
+
+```ts
+and(a: Field, b: Field, length: number) => Field
+```
+
+The bitwise `and()` gadget is a provable equivalent to the [bitwise AND (&)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND) operator in JavaScript. It receives two `Field` elements and compares the corresponding pairs of bits from the binary representation of each. The comparison returns 1 only if both bits are 1 and returns 0 if either bit is not 1. This results in a new binary number which is returned as a `Field` element.
+
+For details about the implementation, see [AND](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#and) in the Mina book.
+
+The `length` parameter:
+
+- Specifies how many bits to compare.
+- Adds more constraints for larger numbers.
+
+Example:
+
+```ts
+let a = Field(3);    // ... 000011
+let b = Field(5);    // ... 000101
+let c = Gadgets.and(a, b, 2);    // ... 000001
+c.assertEquals(1);
+```
+
+## not()
+
+```ts
+not(a: Field, length: number, checked: boolean) => Field
+```
+
+The bitwise `not()` gadget is a provable equivalent to the [Bitwise NOT (~)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_NOT) operator in JavaScript. It receives a `Field` element and negates each bit of its binary representation, turning all the `1`s into `0`s and all the `0`s into `1`s. It essentially flips all the bits in a `Field` element. This results in a new binary number which is returned as a `Field` element.
```

Comment:
so we want to use the tick marks for 1s and 0s here? Seems like we can use plain text 

(love the r/w @nicc !)